### PR TITLE
Add hash-code(string) built-in function

### PR DESCRIPTION
**Pull request by paurue**
The `hash-code(string)` function uses the native java hashCode to generate a hash (int) for a given string.

### CODE REVIEW

**Comment by paurue** [comment]
Good point. My initial approach was to replicate the java's vanilla `hashCode` method, but here it makes more sense to have a hash method that is robust to collisions. Let me give another try...

**Comment by larsga** [comment]
Sure. Internally in Schibsted we already use `sha256-hex`, so if we can use the same name we can get rid of that internal function.

File: src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java [hunk 1]
```diff
@@ -377,21 +379,31 @@ public JsonNode call(JsonNode input, JsonNode[] arguments) {
...
         return arguments[0]; // null
 
-      String string = NodeUtils.toString(arguments[0], false);
-      return new IntNode(string.hashCode());
+      String message = NodeUtils.toString(arguments[0], false);
+      MessageDigest messageDigest;
+      try {
+        messageDigest = MessageDigest.getInstance("SHA-256");
+      } catch (NoSuchAlgorithmException e) {
+        throw new InternalError(e);
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Please throw `JsltException`. We use that for all exceptions thrown by JSLT, to signal that it was JSLT code that failed.

File: src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java [hunk 2]
```diff
@@ -377,21 +379,31 @@ public JsonNode call(JsonNode input, JsonNode[] arguments) {
...
+      }
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Can we get the digest instance once when we create the class, instead of every time we call the function?

File: src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java [hunk 3]
```diff
@@ -80,7 +82,7 @@
     functions.put("join", new BuiltinFunctions.Join());
     functions.put("lowercase", new BuiltinFunctions.Lowercase());
     functions.put("uppercase", new BuiltinFunctions.Uppercase());
-    functions.put("hash-code", new BuiltinFunctions.HashCode());
+    functions.put("sha256", new BuiltinFunctions.Sha256());
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Could we make this `sha256-hex`?

File: src/test/resources/function-tests.json [hunk 4]
```diff
@@ -624,8 +624,8 @@
    },
    {
       "input" : "{}",
-      "output" : "101574",
-      "query" : "hash-code(\"foo\")"
+      "output" : "\"2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae\"",
+      "query" : "sha256(\"foo\")"
    },
```
Comment (larsga) [inline src/test/resources/function-tests.json]: It would be good to also have tests for the `null` case, as well as some integer, and the same integer as a string. If we pass an integer to the function it should be converted to a string and then hashed. The two cases with integers can also be added as examples.

File: src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 1 above)
Comment (paurue) [reply src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Done

File: src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 2 above)
Comment (paurue) [reply src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Ok, done. I am not familiar with java so please review :-)

**Comment by balagueri21** [comment]
@larsga I know this is an old PR post, but can we also add a md5 hash function?

**Comment by larsga** [comment]
It's much better to open new issues for this sort of thing.

I have to say I'm a bit skeptical. If we start down that route we'll soon have a bazillion different hash functions. Writing a Java extension function is easy. Perhaps better to do that?
